tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/Metrics.java

```diff
@@ -49,10 +50,56 @@
    */
   public static <T extends TNumber> Operand<T> topKCategoricalAccuracy(
       Ops tf, Operand<? extends TNumber> labels, Operand<T> predictions, long k) {
-    Operand<TFloat32> fPredictions = CastHelper.cast(tf, predictions, TFloat32.class);
-    return CastHelper.cast(
+    Operand<TFloat32> fPredictions = cast(tf, predictions, TFloat32.class);
+    return cast(
         tf,
         tf.nn.inTopK(fPredictions, tf.math.argMax(labels, tf.constant(-1)), tf.constant(k)),
         predictions.type());
   }
+
+  /**
+   * Computes how often integer targets are in the top K predictions.
+   *
+   * <p>Standalone usage:
+   *
+   * <pre>
+   *     Operand&lt;TInt32&gt; labels = tf.constant(new int[]{2, 1});
+   *     Operand&lt;TFloat32&gt; predictions = tf.constant(new float[][]
+   *                            {{0.1f, 0.9f, 0.f8}, {0.05f, 0.95f, 0f}});
+   *     Operand&lt;TFloat32&gt; m = Metrics.topKCategoricalAccuracy(
+   *                                    labels, predictions, 3)
+   *     //m.shape().toString == "[2]"
+   * </pre>
+   *
+   * @param tf the TensorFlow Ops.
+   * @param labels the ground truth values.
+   * @param predictions The prediction values.
+   * @param k Number of top elements to look at for computing accuracy.
+   * @param <T> the data type for the predictions and results
+   * @param <U> the data type ofr the labels.
+   * @return the Operand for the Sparse top K categorical accuracy value.
+   */
+  @SuppressWarnings("unchecked")
+  public static <T extends TNumber, U extends TNumber> Operand<T> sparseTopKCategoricalAccuracy(
+      Ops tf, Operand<U> labels, Operand<T> predictions, int k) {
+    Operand<T> tLabels = cast(tf, labels, predictions.type());
+
+
+    int predictionsRank = predictions.shape().numDimensions();
+    int labelsRank = tLabels.shape().numDimensions();
+
+    Operand<TFloat32> castPredictions = cast(tf, predictions, TFloat32.class);
+    if (predictionsRank != Shape.UNKNOWN_SIZE && labelsRank != Shape.UNKNOWN_SIZE) {
+      if (predictionsRank > 2) {
+        castPredictions = tf.shape.reduceDims(castPredictions, tf.constant(1));
```

Comment:
If I'm understanding `reduceDims` correctly (which I found very difficult!), then this invocation would have the same result as `flatten`. But what we need is written in Python as:
~~~Python
      y_pred = array_ops.reshape(y_pred, [-1, y_pred.shape[-1]])
~~~

Reply:
I have changed to:


```
//(batch_size, num_samples)
castPredictions = tf.reshape(castPredictions,
                tf.constant(castPredictions.shape().takeLast(1).prepend(Shape.UNKNOWN_SIZE)));
```

and 
//(num_samples)
`tLabels = tf.reshape(tLabels, tf.constant(Shape.of(Shape.UNKNOWN_SIZE));`

Reply:
I am wondering if the `tLabels` reshape should be changed back to `flatten`. In Python, `tf.reshape(inputs, [-1])`, is the same as flatten to 1D. I have tested it with multiple shapes using Python. I need to test this in Java.

Reply:
`tf.reshape(tLabels, tf.constant(Shape.of(Shape.UNKNOWN_SIZE))`
flattens `tLabels`, so this is correct. 

Could also be:` tf.reshape(tLabels, tf.constant(new long[] { -1L}))`